# A Simple Differential Gene Expression Analysis

# Package installations if needed 
```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("DESeq2")
install.packages("tidyverse")
install.packages("tximport")
BiocManager::install("EnsDb.Hsapiens.v86") # This package loads an SQL connection to a database containing annotations from Ensembl
BiocManager::install('EnhancedVolcano') # highly-configurable publication-ready volcano plots.
```

The script are adapted from <https://diytranscriptomics.com>.

The dataset is an RNAseq dataset from skin biopsies obtained from patients with cutaneous leishmaniasis, a parasitic disease endemic in Brazil and other areas of South America. These datasets were obtained also from <https://diytranscriptomics.com>.

We'll be working with data from 5 patients with this disease and 5 healthy endemic controls.

## DGE Analysis objective!



# Step 1: Prepare the count data

```{r}
library(DESeq2) # This package is used for differential gene expression analysis
library(tidyverse) # For data wrangling
library(tximport) # Transcript related things
library(ensembldb) # database from ENSEMBL
library(EnsDb.Hsapiens.v86) # This package loads an SQL connection to a database containing annotations from Ensembl
library(EnhancedVolcano) # highly-configurable publication-ready volcano plots.
```

```{r}
# Read the study design and pat
studyDesign <- read_delim("data/studydesign.txt") # read in your study design
path <- file.path("data", "kallisto", studyDesign$sample, "abundance.tsv") # set file paths to your mapped data

# Check if the path exist
all(file.exists(path))

# Check the study design
studyDesign
```

```{r}
# Prepare transcript to gene annotation using ensemble
Tx <- transcripts(EnsDb.Hsapiens.v86, columns=c("tx_id", "gene_name"))
Tx <- as_tibble(Tx)
Tx <- dplyr::rename(Tx, target_id = tx_id)
Tx <- dplyr::select(Tx, "target_id", "gene_name")

# Load kallisto data
Txi_gene <- tximport(
  path, 
  type = "kallisto", 
  tx2gene = Tx, 
  txOut = FALSE, #determines whether your data represented at transcript or gene level
  countsFromAbundance = "no",
  ignoreTxVersion = TRUE
)
```

### Take a glance to our data

```{r}
# Make a counts dataframe from Txi_gene
counts <- Txi_gene$counts
counts <- as.data.frame(counts, rownames = "geneID")
sampleLabels <- studyDesign$sample
colnames(counts) <- c(sampleLabels)

head(counts)
dim(counts)
```

# Step 2: Differential expression analysis using DESeq2

## Add Txi to dds object

```{r}
# We use DESeqDataSetFromTximport because we use tximport to load the data
dds <- DESeqDataSetFromTximport(txi = Txi_gene, 
                                colData = studyDesign, 
                                design = ~ group)
```

## Filter data

```{r}
groupSize <- 5
keep <- rowSums(counts(dds) >= 10) >= groupSize
dds <- dds[keep, ]
dds$group <- relevel(dds$group, ref = "healthy")

dds
```

## Perform DESeq2

```{r}
dds <- DESeq(dds)
```

```{r}
resultsNames(dds)
```

## Get the result

```{r}
# Get the result
res <- results(dds)
```

## Explore the result

```{r}
summary(res)
```

## More filtering

```{r}

# Set this value in your favor
ALPHA = 0.01
LFCTHRESHOLD = 0.25

# Get the result
res <- results(dds, alpha=ALPHA, lfcThreshold = LFCTHRESHOLD)
summary(res)
```

## Log fold shrinkage

```{r}
# Log fold change shrinkage for visualization and ranking
resLFC <- lfcShrink(dds, coef=2, res=res) 
```

```{r}
resLFC
```

```{r}
summary(resLFC)
```

# Step 4: Data Analysis

```{r}
vsd <- vst(dds, blind = FALSE)
data <- assay(vsd)
```

## Principal Component Analysis

```{r}
# Perform PCA
pca_result <- prcomp(t(data))
pca_exp_var <- summary(pca_result)$importance["Proportion of Variance",]
# Create a data frame with the PCA results
pca_df <- as.data.frame(pca_result$x)
rownames(pca_df) <- c(sampleLabels)
group <- studyDesign$group
group <- factor(group)



# Create the PCA plot
ggplot(pca_df, aes(x = PC1, y = PC2, color = group)) +
  geom_point() +
  geom_text(aes(label = rownames(pca_df)), vjust = -0.5, show.legend = FALSE) +
  theme_minimal() +
  labs(title = "PCA of Gene Expression Samples",
       x = paste0("Principal Component 1 (", pca_exp_var[1]*100, "%)"),
       y = paste0("Principal Component 2 (", pca_exp_var[2]*100, "%)"))
```

## Hierarchical Clustering

```{r}
distance <- dist(t(data), method = "euclidean") 
# other distance methods are "euclidean", maximum", "manhattan", "canberra", "binary" or "minkowski"

clusters <- hclust(distance, method = "average") 
# other agglomeration methods are "ward.D", "ward.D2", "single", "complete", "average", "mcquitty", "median", or "centroid"

plot(clusters, labels=sampleLabels)
```

## Top hits

```{r}
# Sort by adjusted p-value
res_sorted <- resLFC[order(resLFC$padj),]

# Filter for significant genes
top_hits <- res_sorted[which(res_sorted$padj < 0.05),]

# Add a column for absolute log2FoldChange
top_hits$absLFC <- abs(top_hits$log2FoldChange)

# Sort by a combination of significance and effect size
top_hits <- top_hits[order(top_hits$padj * (1 / top_hits$absLFC)),]

# Display top genes
head(top_hits, n = 20)
```

## MA plot

```{r}
ggplot(resLFC, mapping = aes(x = baseMean, y = log2FoldChange,
                             color=padj < ALPHA)) +
  geom_point(alpha=0.4) +
  geom_hline(yintercept = 0, alpha = 0.75, color = "red") +
  geom_hline(yintercept = -1) +
  geom_hline(yintercept = 1) +
  geom_smooth(colour = "blue") +
  scale_x_log10() +
  labs(title = "MA Plots") +
  theme_bw()
```


## Volcano plot

<https://bioconductor.org/packages/devel/bioc/vignettes/EnhancedVolcano/inst/doc/EnhancedVolcano.html>

```{r}
EnhancedVolcano(resLFC,
    lab = rownames(resLFC),
    x = 'log2FoldChange',
    y = 'padj')
```

```{r}
plotCounts(dds, gene=which.min(resLFC$padj), intgroup="group")
```
